R/sampler-adaptive.R

```diff
@@ -0,0 +1,310 @@
+##'   better, but while the chains find the location and scale of the posterior
+##'   mode it might be useful to explore with it switched off.
+##'
+##'
+##' @return A `mcstate_sampler` object, which can be used with
+##'   [mcstate_sample]
+##'
+##' @export
+mcstate_sampler_adaptive <- function(initial_vcv,
+                                     initial_vcv_weight = 1000,
+                                     initial_scaling = 1,
+                                     min_scaling = 0,
+                                     scaling_increment = NULL,
+                                     log_scaling_update = TRUE,
+                                     acceptance_target = 0.234,
+                                     forget_rate = 0.2,
+                                     forget_end = Inf,
+                                     adapt_end = Inf,
+                                     pre_diminish = 0) {
+  ## This sampler is stateful; we will be updating our estimate of the
+  ## mean and vcv of the target distribution, along with the our
+  ## scaling factor, weight and autocorrelations.
+  ##
+  ## Probably we will provide some method later for extracting this
+  ## internal state and dump it out at the end of the sampling (so
+  ## perhaps 'finalise') as I think we've previously collected this up
+  ## at the end of the sampling as the estimated vcv is of interest.
+  internal <- new.env()
+
+  initialise <- function(state, model, rng) {
+    internal$weight <- 0
+    internal$iteration <- 0
+    
+    internal$mean <- state$pars
+    d <- length(internal$mean)
+    internal$autocorrelation <- matrix(0, d, d)
+    internal$vcv <- update_vcv(internal$mean, internal$autocorrelation,
+                               internal$weight)
+    
+    internal$scaling <- initial_scaling
+    internal$scaling_increment <- scaling_increment %||%
+      calc_scaling_increment(length(internal$mean), acceptance_target,
+                             log_scaling_update)
+    internal$n_start <- calc_n_start(acceptance_target)
+    
+    internal$history_pars <- c()
+    internal$included <- c()
+    internal$scaling_history <- internal$scaling
+  }
+
+  step <- function(state, model, rng) {
+    proposal_vcv <- 
+      calc_proposal_vcv(internal$scaling, internal$vcv, internal$weight,
+                        initial_vcv, initial_vcv_weight)
+    
+    pars_next <- rmvnorm(state$pars, proposal_vcv, rng)
+
+    u <- rng$random_real(1)
+    density_next <- model$density(pars_next)
+    
+    accept_prob <- min(1, exp(density_next - state$density))
+
+    accept <- u < accept_prob
+    if (accept) {
+      state$pars <- pars_next
+      state$density <- density_next
+    }
+
+    internal$iteration <- internal$iteration + 1
+    internal$history_pars <- rbind(internal$history_pars, state$pars) 
+    if (internal$iteration > adapt_end) {
+      internal$scaling_history <- c(internal$scaling_history, internal$scaling)
```

Comment:
Does this mean that `internal$scaling_history` has length `internal$iteration`? If so can we use this as the source of truth for iteration?

Reply:
Yes, but also the length of `internal$scaling_history` could be pre-allocated (like you suggest with `internal$history_pars`).